controllers/routing/routing_controller_test.go

```diff
@@ -0,0 +1,578 @@
+					), "public services are not expected to be defined in this mode")
+
+				g.Expect(updatedComponent).To(
+					HaveAnnotations(
+						metadata.Annotations.RoutingAddressesPublic,
+						fmt.Sprintf("%[1]s-%[2]s.%[3]s;%[1]s-%[2]s.%[3]s.svc;%[1]s-%[2]s.%[3]s.svc.cluster.local", svc.Name, svc.Namespace, routingConfiguration.GatewayNamespace),
+					),
+				)
+
+				return nil
+			}).
+				WithContext(ctx).
+				WithTimeout(test.DefaultTimeout).
+				WithPolling(test.DefaultPolling).
+				Should(Succeed())
+		})
+
+	})
+
+	When("component requests to expose service both locally and externally to the cluster", func() {
+
+		It("should have both external and cluster-local resources created", func(ctx context.Context) {
+			// given
+			component, errCreate := test.CreateResource(ctx, envTest.Client,
+				componentResource("public-and-external-test-component", appNs.Name))
+			Expect(errCreate).ToNot(HaveOccurred())
+			toRemove = append(toRemove, component)
+
+			// when
+			By("adding routing requirements on the resource and related svc", func() {
+				// routing.opendatahub.io/exported: "true"
+				exportAnnotation := metadata.WithLabels(metadata.Labels.RoutingExported, "true")
+				// platform.opendatahub.io/owner-name: test-component
+				// platform.opendatahub.io/owner-kind: Component
+				ownerLabels := metadata.WithOwnerLabels(component)
+
+				// Service created by the component need to have these metadata added, i.e. by its controller
+				_, errExportSvc := controllerutil.CreateOrUpdate(ctx, envTest.Client, svc, func() error {
+					return metadata.ApplyMetaOptions(svc, exportAnnotation, ownerLabels)
+				})
+				Expect(errExportSvc).ToNot(HaveOccurred())
+
+				// routing.opendatahub.io/export-mode: "public;external"
+				exposeExternally := metadata.WithAnnotations(metadata.Annotations.RoutingExportMode, "public;external")
+				_, errExportCR := controllerutil.CreateOrUpdate(
+					ctx, envTest.Client,
+					component,
+					func() error {
+						return metadata.ApplyMetaOptions(component, exposeExternally)
+					})
+				Expect(errExportCR).ToNot(HaveOccurred())
+			})
+
+			// then
+			Eventually(routeExistsFor(svc)).
+				WithContext(ctx).
+				WithTimeout(test.DefaultTimeout).
+				WithPolling(test.DefaultPolling).
+				Should(Succeed())
+
+			Eventually(ingressVirtualServiceExistsFor(svc)).
+				WithContext(ctx).
+				WithTimeout(test.DefaultTimeout).
+				WithPolling(test.DefaultPolling).
+				Should(Succeed())
+
+			Eventually(publicSvcExistsFor(svc)).
+				WithContext(ctx).
+				WithTimeout(test.DefaultTimeout).
+				WithPolling(test.DefaultPolling).
+				Should(Succeed())
+
+			Eventually(publicVirtualSvcExistsFor(svc)).
+				WithContext(ctx).
+				WithTimeout(test.DefaultTimeout).
+				WithPolling(test.DefaultPolling).
+				Should(Succeed())
+
+			Eventually(publicGatewayExistsFor(svc)).
+				WithContext(ctx).
+				WithTimeout(test.DefaultTimeout).
+				WithPolling(test.DefaultPolling).
+				Should(Succeed())
+
```

Comment:
worth adding one for destination rule too? and above in the public only test?

Reply:
Added in d5da2e6